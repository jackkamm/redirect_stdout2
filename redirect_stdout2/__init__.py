from contextlib import contextmanager
import sys
import traceback

@contextmanager
def redirect_stdout(stream, stdout=True, stderr=False):
    if stdout:
        sys.stdout = stream
    if stderr:
        sys.stderr = stream
    try:
        yield
    except Exception,e:
        if stderr:
            traceback.print_exc()
        raise
    finally:
        if stdout:
            sys.stdout = sys.__stdout__
        if stderr:
            sys.stderr = sys.__stderr__

def redirect_stderr(stream):
    return redirect_stdout(stream, stdout=False, stderr=True)
