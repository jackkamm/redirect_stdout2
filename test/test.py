from redirect_stdout2 import redirect_stdout, redirect_stderr

with file('test.out','w') as f:
    with redirect_stdout(f):
        print "foo"
    with redirect_stdout(f, stderr=True, stdout=False):
        raise Exception("bar")


