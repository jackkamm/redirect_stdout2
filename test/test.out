foo
Traceback (most recent call last):
  File "/home/snackattack/anaconda/lib/python2.7/site-packages/redirect_stdout2/__init__.py", line 12, in redirect_stdout
    yield
  File "test.py", line 7, in <module>
    raise Exception("bar")
Exception: bar
